--- /usr/lib/scons/SCons/Util.py	2008-12-20 23:59:59.000000000 -0700
+++ scons/SCons/Util.py	2009-11-04 15:54:31.000000000 -0700
@@ -640,6 +640,13 @@
     if copier:
         return copier(x)
     else:
+        if hasattr(x, '__semi_deepcopy__'):
+            return x.__semi_deepcopy__()
+        elif isinstance(x, UserDict):
+            return x.__class__(_semi_deepcopy_dict(x))
+        elif isinstance(x, UserList):
+            return x.__class__(_semi_deepcopy_list(x))
+        
         return x
 
 
