--- setup.py    2008-06-09 22:21:08.000000000 -0700
+++ ../scons-0.98.5-patched/setup.py    2008-06-20 11:33:11.701172000 -0700
@@ -361,2 +361,2 @@
         else:
             self.data_files = []

-description = """Open Source next-generation build tool.
+description = "Open Source next-generation build tool."
+
+long_description = """Open Source next-generation build tool.
 Improved, cross-platform substitute for the classic Make
 utility.  In short, SCons is an easier, more reliable
 and faster way to build software."""
@@ -387,6 +389,7 @@
     'name'             : "scons",
     'version'          : Version,
     'description'      : description,
+    'long_description' : long_description,
     'author'           : 'Steven Knight',
     'author_email'     : 'knight@baldmt.com',
     'url'              : "http://www.scons.org/",
