Index: Util.py
===================================================================
RCS file: /cvs/scons/src/engine/SCons/Util.py,v
retrieving revision 1.122
diff -u -r1.122 Util.py
--- Util.py	10 Jan 2005 13:44:58 -0000	1.122
+++ Util.py	18 Jun 2005 01:02:07 -0000
@@ -546,11 +546,12 @@
 #
 #       "   "                   [white space]
 #       "non-white-space"       [without any dollar signs]
+#       '"with-space-or-not"'   [without any dollar signs]
 #       "$"                     [single dollar sign]
 #
 _dollar_exps_str = r'\$[\$\(\)]|\$[_a-zA-Z][\.\w]*|\${[^}]*}'
 _dollar_exps = re.compile(r'(%s)' % _dollar_exps_str)
-_separate_args = re.compile(r'(%s|\s+|[^\s\$]+|\$)' % _dollar_exps_str)
+_separate_args = re.compile(r'(%s|\s+|[^"\s\$]+|"[^"\$]+"|\$)' % _dollar_exps_str)
 
 # This regular expression is used to replace strings of multiple white
 # space characters in the string result from the scons_subst() function.
@@ -879,6 +880,8 @@
                         else:
                             self.next_word()
                     else:
+                        if a[0] == '"' and a[-1] == '"':
+                            a = a[1:-1]
                         self.expand(a, lvars, within_list)
             else:
                 self.expand(args, lvars, within_list)
