--- __init__.py~	Mon Apr 10 22:33:21 2006
+++ __init__.py	Wed Oct 11 19:17:28 2006
@@ -536,7 +536,7 @@
         # Here's where we implement --implicit-cache.
         if implicit_cache and not implicit_deps_changed:
             implicit = self.get_stored_implicit()
-            if implicit:
+            if implicit is not None:
                 factory = build_env.get_factory(self.builder.source_factory)
                 nodes = []
                 for i in implicit:
