Index: zip.py
===================================================================
--- zip.py	(revision 5081)
+++ zip.py	(working copy)
@@ -55,12 +55,12 @@
                 if os.path.isfile(path):
                     arg.write(path)
         compression = env.get('ZIPCOMPRESSION', 0)
-        zf = zipfile.ZipFile(str(target[0]), 'w', compression)
+        zf = zipfile.ZipFile(target[0].abspath, 'w', compression)
         for s in source:
             if s.isdir():
-                os.path.walk(str(s), visit, zf)
+                os.path.walk(os.path.relpath(s.abspath), visit, zf)
             else:
-                zf.write(str(s))
+                zf.write(os.path.relpath(s.abspath))
         zf.close()
 else:
     zipcompression = 0
