--- mingw.py.orig	2011-11-06 13:36:40 +0200
+++ mingw.py	2011-11-06 13:37:25 +0200
@@ -117,7 +117,8 @@
     return (target, source)
                          
 
-shlib_action = SCons.Action.Action(shlib_generator, generator=1)
+shlink_action = SCons.Action.Action(shlib_generator, '$SHLINKCOMSTR', generator=1)
+ldmodule_action = SCons.Action.Action(shlib_generator, '$LDMODULECOMSTR', generator=1)
 
 res_action = SCons.Action.Action('$RCCOM', '$RCCOMSTR')
 
@@ -143,8 +144,8 @@
     env['CXX'] = 'g++'
     env['SHCXXFLAGS'] = SCons.Util.CLVar('$CXXFLAGS')
     env['SHLINKFLAGS'] = SCons.Util.CLVar('$LINKFLAGS -shared')
-    env['SHLINKCOM']   = shlib_action
-    env['LDMODULECOM'] = shlib_action
+    env['SHLINKCOM']   = shlink_action
+    env['LDMODULECOM'] = ldmodule_action
     env.Append(SHLIBEMITTER = [shlib_emitter])
     env['AS'] = 'as'
 
