diff -r 023b335b70db src/engine/SCons/Environment.py
--- a/src/engine/SCons/Environment.py	Sun Aug 31 20:50:05 2014 +0200
+++ b/src/engine/SCons/Environment.py	Sun Aug 31 23:30:17 2014 +0200
@@ -1854,7 +1854,8 @@
         action = SCons.Action.Action(action)
         uniq = {}
         for executor in [n.get_executor() for n in nodes]:
-            uniq[executor] = 1
+            if executor is not None:
+                uniq[executor] = 1
         for executor in uniq.keys():
             executor.add_pre_action(action)
         return nodes
@@ -1864,7 +1865,8 @@
         action = SCons.Action.Action(action)
         uniq = {}
         for executor in [n.get_executor() for n in nodes]:
-            uniq[executor] = 1
+            if executor is not None:
+                uniq[executor] = 1
         for executor in uniq.keys():
             executor.add_post_action(action)
         return nodes
diff -r 023b335b70db src/engine/SCons/Taskmaster.py
--- a/src/engine/SCons/Taskmaster.py	Sun Aug 31 20:50:05 2014 +0200
+++ b/src/engine/SCons/Taskmaster.py	Sun Aug 31 23:30:17 2014 +0200
@@ -803,7 +803,10 @@
             executor = node.get_executor()
 
             try:
-                children = executor.get_all_children()
+                if executor is not None:
+                    children = executor.get_all_children()
+                else:
+                    children = []
             except SystemExit:
                 exc_value = sys.exc_info()[1]
                 e = SCons.Errors.ExplicitExit(node, exc_value.code)
