--- scons-0.96.95/engine/SCons/Environment.py	2007-02-12 19:54:58.000000000 -0800
+++ scons-0.96.95.patched/engine/SCons/Environment.py	2007-03-15 20:59:33.000000000 -0700
@@ -1046,14 +1046,19 @@
 
         clone._memo = {}
 
-        apply_tools(clone, tools, toolpath)
-
-        # Apply passed-in variables after the new tools.
+        # Apply passed-in variables before the tools
+        # so the tools can use the new variables
         kw = copy_non_reserved_keywords(kw)
         new = {}
         for key, value in kw.items():
             new[key] = SCons.Subst.scons_subst_once(value, self, key)
         apply(clone.Replace, (), new)
+
+        apply_tools(clone, tools, toolpath)
+
+        # apply them again in case the tools overwrote them
+        apply(clone.Replace, (), new)        
+
         if __debug__: logInstanceCreation(self, 'Environment.EnvironmentClone')
         return clone
 
