--- a/Script/Main.py    Wed Feb 27 14:38:58 2013 +0100
+++ b/Script/Main.py    Wed Feb 27 15:14:29 2013 +0100
@@ -407,13 +407,28 @@ class TreePrinter(object):
     def get_derived_children(self, node):
         children = node.all_children(None)
         return [x for x in children if x.has_builder()]
+    def get_min_depth(self, root, margin=[0], min_depth={}):
+        rname = str(root)
+        min_depth[rname] = min(min_depth.get(rname, sys.maxint), len(margin[:-1
]))
+        children = root.all_children()
+        if children:
+            margin.append(1)
+            for C in children[:-1]:
+                self.get_min_depth(C, margin, min_depth)
+            margin[-1] = 0
+            self.get_min_depth(children[-1], margin, min_depth)
+            margin.pop()
+
+        return min_depth
+
     def display(self, t):
         if self.derived:
             func = self.get_derived_children
         else:
             func = self.get_all_children
         s = self.status and 2 or 0
-        SCons.Util.print_tree(t, func, prune=self.prune, showtags=s)
+        min_depth = self.get_min_depth(t)
+        SCons.Util.print_tree(t, func, min_depth, prune=self.prune, showtags=s)



 def python_version_string():