Index: src/engine/SCons/Platform/win32.py
===================================================================
RCS file: /cvs/scons/src/engine/SCons/Platform/win32.py,v
retrieving revision 1.30
diff -u -r1.30 win32.py
--- src/engine/SCons/Platform/win32.py	24 Oct 2004 14:29:53 -0000	1.30
+++ src/engine/SCons/Platform/win32.py	9 Feb 2005 21:47:52 -0000
@@ -193,7 +193,13 @@
 
 # Windows does not allow special characters in file names anyway, so
 # no need for a complex escape function, we will just quote the arg.
-escape = lambda x: '"' + x + '"'
+def escape(x):
+    # There is a special case with 'cmd /c' when if argument ends with
+    # a backslash it needs to be escaped so that not to interfere with
+    # closing double quote that we add here.
+    if x[-1] == '\\':
+        x += '\\'
+    return '"' + x + '"'
 
 # Get the windows system directory name
 def get_system_root():
