Index: src/engine/SCons/Tool/msvc.py
===================================================================
--- src/engine/SCons/Tool/msvc.py	(revision 5214)
+++ src/engine/SCons/Tool/msvc.py	(working copy)
@@ -140,8 +140,10 @@
     Returning None specifies that the specified target+source should not
     be batched with other compilations.
     """
-    b = env.subst('$MSVC_BATCH')
-    if b in (None, '', '0'):
+
+    # Fixing MSVC_BATCH mode. Previous if did not work when MSVC_BATCH
+    # was set to False. This new version should work better.
+    if not 'MSVC_BATCH' in env or not env['MSVC_BATCH']:
         # We're not using batching; return no key.
         return None
     t = target[0]
@@ -161,8 +163,13 @@
     we return an /Fo string that just specifies the first target's
     directory (where the Visual C/C++ compiler will put the .obj files).
     """
-    b = env.subst('$MSVC_BATCH')
-    if b in (None, '', '0') or len(source) == 1:
+
+    # Fixing MSVC_BATCH mode. Previous if did not work when MSVC_BATCH
+    # was set to False. This new version should work better. Removed
+    # len(source)==1 as batch mode can compile only one file
+    # (and it also fixed problem with compiling only one changed file
+    # with batch mode enabled)
+    if not 'MSVC_BATCH' in env or not env['MSVC_BATCH']:
         return '/Fo$TARGET'
     else:
         # The Visual C/C++ compiler requires a \ at the end of the /Fo
