Index: src/engine/SCons/Tool/dmd.py
===================================================================
--- src/engine/SCons/Tool/dmd.py	(revision 4084)
+++ src/engine/SCons/Tool/dmd.py	(working copy)
@@ -200,7 +200,7 @@
                         libs = env['LIBS']
                     except KeyError:
                         libs = []
-                    if 'phobos' not in libs:
+                    if 'phobos' not in libs and 'gphobos' not in libs:
                         if dc is 'dmd':
                             env.Append(LIBS = ['phobos'])
                         elif dc is 'gdmd':
