diff -Naur scons-src-0.96.95/src/engine/SCons/Taskmaster.py scons-src-0.96.95-dup_kids/src/engine/SCons/Taskmaster.py
--- scons-src-0.96.95/src/engine/SCons/Taskmaster.py	2007-02-12 19:56:15.000000000 -0800
+++ scons-src-0.96.95-dup_kids/src/engine/SCons/Taskmaster.py	2007-03-28 21:40:15.000000000 -0700
@@ -103,6 +103,14 @@
     for n in StatsNodes:
         print (fmt % n.stats.__dict__) + str(n)
 
+def count_unique_nodes(nodes):
+    node_dict = {}
+    num_unique = 0
+    for n in nodes:
+        if not node_dict.has_key(n):
+            node_dict[n] = 1
+            num_unique = num_unique + 1
+    return num_unique
 
 
 class Task:
@@ -572,7 +580,7 @@
                 # list will get cleared and we'll re-scan the newly-built
                 # file(s) for updated implicit dependencies.
                 map(lambda n, P=node: n.add_to_waiting_parents(P), not_started)
-                node.ref_count = len(not_started)
+                node.ref_count = count_unique_nodes(not_started)
 
                 # Now we add these derived targets to the candidates
                 # list so they can be examined and built.  We have to
@@ -607,7 +615,7 @@
                 # dependency list will get cleared and we'll re-scan the
                 # newly-built file(s) for updated implicit dependencies.
                 map(lambda n, P=node: n.add_to_waiting_parents(P), not_built)
-                node.ref_count = len(not_built)
+                node.ref_count = count_unique_nodes(not_built)
 
                 if S: S.not_built = S.not_built + 1
                 if T:
