Index: src/engine/SCons/Node/FS.py
===================================================================
--- src/engine/SCons/Node/FS.py	(revision 3271)
+++ src/engine/SCons/Node/FS.py	(working copy)
@@ -1593,9 +1593,12 @@
         return None
 
     def get_contents(self):
-        """Return aggregate contents of all our children."""
-        contents = map(lambda n: n.get_contents(), self.children())
-        return  string.join(contents, '')
+        """Return content signatures and names of all our children separated by
+        new-lines."""
+        contents = []
+        for node in self.children():
+            contents.append('%s %s\n' % (node.get_csig(), node.name))
+        return string.join(contents, '')        
 
     def do_duplicate(self, src):
         pass
Index: src/engine/SCons/Node/FSTests.py
===================================================================
--- src/engine/SCons/Node/FSTests.py	(revision 3271)
+++ src/engine/SCons/Node/FSTests.py	(working copy)
@@ -1689,6 +1689,23 @@
                         os.path.join('ddd', 'f2'),
                         os.path.join('ddd', 'f3')], kids
 
+    def test_get_contents(self):
+        """Test getting the contents for a directory.
+        """
+        test = self.test
+
+        test.subdir('d')
+        test.write(['d', 'f'], "12345\n")
+        test.write(['d', 'g'], "67890\n")
+
+        d = self.fs.Dir('d')
+        f = self.fs.File(os.path.join('d', 'f'))
+        g = self.fs.File(os.path.join('d', 'g'))
+        files = d.get_contents().split('\n')
+
+        assert f.get_csig()+" f" in files, files
+        assert g.get_csig()+" g" in files, files
+
     def test_implicit_re_scans(self):
         """Test that adding entries causes a directory to be re-scanned
         """
