Index: test/WhereIs.py
===================================================================
--- test/WhereIs.py	(revision 3285)
+++ test/WhereIs.py	(working copy)
@@ -74,13 +74,16 @@
 print WhereIs('xxx.exe', %s)
 print WhereIs('xxx.exe', %s)
 print WhereIs('xxx.exe', %s, reject=%s)
+env.Replace( XXXNAME='xxx.exe' )
+print env.WhereIs( '$XXXNAME', %s )
 """ % (subdir_SConscript,
        repr(string.join(pathdirs_1234, os.pathsep)),
        repr(string.join(pathdirs_1243, os.pathsep)),
        repr(pathdirs_1234),
        repr(pathdirs_1243),
        repr(pathdirs_1243),
-       repr(sub4_xxx_exe)
+       repr(sub4_xxx_exe),
+       repr(string.join(pathdirs_1243, os.pathsep)),
       ))
 
 test.write(subdir_SConscript, """
@@ -109,6 +112,7 @@
            test.workpath(sub3_xxx_exe),
            test.workpath(sub4_xxx_exe),
            test.workpath(sub3_xxx_exe),
+           test.workpath(sub4_xxx_exe)
          ]
 
 test.run(arguments = ".",
@@ -128,10 +132,12 @@
            test.workpath(sub3_xxx_exe),
            test.workpath(sub4_xxx_exe),
            test.workpath(sub3_xxx_exe),
+           test.workpath(sub4_xxx_exe)
          ]
 
 test.run(arguments = ".",
          stdout = test.wrap_stdout(read_str = string.join(expect, "\n") + "\n",
                                    build_str = "scons: `.' is up to date.\n"))
 
+
 test.pass_test()
Index: src/engine/SCons/Environment.py
===================================================================
--- src/engine/SCons/Environment.py	(revision 3285)
+++ src/engine/SCons/Environment.py	(working copy)
@@ -1599,6 +1599,7 @@
                 pass
         elif SCons.Util.is_String(pathext):
             pathext = self.subst(pathext)
+        prog = self.subst(prog)
         path = SCons.Util.WhereIs(prog, path, pathext, reject)
         if path: return path
         return None
