--- scons-3.0.0-orig/SCons/Node/FS.py	2017-09-18 13:59:26.000000000 +0200
+++ scons-3.0.0/SCons/Node/FS.py	2017-11-01 16:00:20.285400000 +0100
@@ -59,6 +59,14 @@

 print_duplicate = 0

+# work around an issue with python on windows where shutil.copy<2> methods
+# didn't appropriately close files, which caused build failures with the cached build.
+if sys.platform == "win32":
+    import win32api
+    _copy = _copy2 = win32api.CopyFile
+else:
+    _copy = shutil.copy
+    _copy2 = shutil.copy2

 def sconsign_none(node):
     raise NotImplementedError
@@ -248,7 +256,7 @@
     _softlink_func = None

 def _copy_func(fs, src, dest):
-    shutil.copy2(src, dest)
+    _copy2(src, dest)
     st = fs.stat(src)
     fs.chmod(dest, stat.S_IMODE(st[stat.ST_MODE]) | stat.S_IWRITE)

@@ -1068,9 +1076,9 @@
     def chmod(self, path, mode):
         return os.chmod(path, mode)
     def copy(self, src, dst):
-        return shutil.copy(src, dst)
+        return _copy(src, dst)
     def copy2(self, src, dst):
-        return shutil.copy2(src, dst)
+        return _copy2(src, dst)
     def exists(self, path):
         return os.path.exists(path)
     def getmtime(self, path):
