diff -r 05db74dca43c src/engine/SCons/Tool/__init__.py
--- a/src/engine/SCons/Tool/__init__.py	Sun Mar 03 19:34:10 2013 -0500
+++ b/src/engine/SCons/Tool/__init__.py	Mon Sep 09 19:38:58 2013 +0200
@@ -257,6 +257,10 @@
             print "VersionShLibLinkNames: linkname = ",linkname
         linknames.append(linkname)
     elif platform == 'posix':
+        if sys.platform.startswith('openbsd'):
+            # OpenBSD uses x.y shared library versioning numbering convention
+            # and doesn't use symlinks to backwards-compatible libraries
+            return []
         # For libfoo.so.x.y.z, linknames libfoo.so libfoo.so.x.y libfoo.so.x
         suffix_re = re.escape(shlib_suffix + '.' + version)
         # First linkname has no version number
@@ -302,13 +306,17 @@
     if version:
         # set the shared library link flags
         if platform == 'posix':
-            suffix_re = re.escape(shlib_suffix + '.' + version)
-            (major, age, revision) = version.split(".")
-            # soname will have only the major version number in it
-            soname = re.sub(suffix_re, shlib_suffix, libname) + '.' + major
-            shlink_flags += [ '-Wl,-Bsymbolic', '-Wl,-soname=%s' % soname ]
-            if Verbose:
-                print " soname ",soname,", shlink_flags ",shlink_flags
+            shlink_flags += [ '-Wl,-Bsymbolic' ]
+            if sys.platform.startswith('openbsd'):
+                pass # OpenBSD doesn't usually use SONAME for libraries
+            else:
+                suffix_re = re.escape(shlib_suffix + '.' + version)
+                (major, age, revision) = version.split(".")
+                # soname will have only the major version number in it
+                soname = re.sub(suffix_re, shlib_suffix, libname) + '.' + major
+                shlink_flags += [ '-Wl,-soname=%s' % soname ]
+                if Verbose:
+                    print " soname ",soname,", shlink_flags ",shlink_flags
         elif platform == 'cygwin':
             shlink_flags += [ '-Wl,-Bsymbolic',
                               '-Wl,--out-implib,${TARGET.base}.a' ]
diff -r 05db74dca43c test/Libs/SharedLibrary.py
--- a/test/Libs/SharedLibrary.py	Sun Mar 03 19:34:10 2013 -0500
+++ b/test/Libs/SharedLibrary.py	Mon Sep 09 19:38:58 2013 +0200
@@ -60,6 +60,13 @@
 Default(env.Library(target = 'foo', source = obj))
 """)
 
+test.write('SConstructBaz', """
+env=Environment()
+env['SHLIBVERSION'] = '1.0.0'
+obj = env.SharedObject('baz', 'foo.c')
+Default(env.SharedLibrary(target = 'baz', source = obj))
+""")
+
 test.write('foo.c', r"""
 #include <stdio.h>
 
@@ -280,6 +287,12 @@
     test.run(program = test.workpath('progbar'),
              stdout = "f4.c\nprogbar.c\n")
 
+if sys.platform.startswith('openbsd'):
+    # Make sure we don't link libraries with -Wl,-soname on OpenBSD.
+    test.run(arguments = '-f SConstructBaz')
+    for line in test.stdout().split('\n'):
+        test.fail_test(line.find('-Wl,-soname=libbaz.so') != -1)
+
 test.pass_test()
 
 # Local Variables:

