Index: src/engine/SCons/Script/Main.py
===================================================================
--- src/engine/SCons/Script/Main.py	(revision 2660)
+++ src/engine/SCons/Script/Main.py	(working copy)
@@ -1184,7 +1184,10 @@
         import pdb
         pdb.Pdb().runcall(_main, parser)
     elif options.profile_file:
-        from profile import Profile
+        try:
+            from cProfile import Profile
+        except ImportError, e:
+            from profile import Profile
 
         # Some versions of Python 2.4 shipped a profiler that had the
         # wrong 'c_exception' entry in its dispatch table.  Make sure
