--- Fortran.py	2009-09-20 01:00:01.000000000 +0200
+++ Fortran 2.py	2009-10-29 10:35:30.000000000 +0100
@@ -90,6 +90,10 @@
             # retrieve all defined module names
             defmodules = self.cre_def.findall(node.get_text_contents())
 
+            # convert module names to lower case as fortran is not case sensitive
+            modules = [string.lower(mod) for mod in modules]
+            defmodules = [string.lower(mod) for mod in defmodules]
+
             # Remove all USE'd module names that are defined in the same file
             d = {}
             for m in defmodules:
