Index: src/engine/SCons/Node/__init__.py
===================================================================
--- src/engine/SCons/Node/__init__.py	(revision 5354)
+++ src/engine/SCons/Node/__init__.py	(working copy)
@@ -216,6 +216,7 @@
         self.precious = None
         self.noclean = 0
         self.nocache = 0
+        self.cached = 0 # is this node pulled from cache?
         self.always_build = None
         self.includes = None
         self.attributes = self.Attrs() # Generic place to stick information about the Node.
Index: src/engine/SCons/TaskmasterTests.py
===================================================================
--- src/engine/SCons/TaskmasterTests.py	(revision 5354)
+++ src/engine/SCons/TaskmasterTests.py	(working copy)
@@ -91,6 +91,9 @@
         global built_text
         built_text = self.name + " built"
 
+    def remove(self):
+        pass
+
     def built(self):
         global built_text
         if not self.cached:
Index: src/engine/SCons/Script/Main.py
===================================================================
--- src/engine/SCons/Script/Main.py	(revision 5354)
+++ src/engine/SCons/Script/Main.py	(working copy)
@@ -233,6 +233,7 @@
         # Handle the failure of a build task.  The primary purpose here
         # is to display the various types of Errors and Exceptions
         # appropriately.
+
         exc_info = self.exc_info()
         try:
             t, e, tb = exc_info
@@ -240,6 +241,12 @@
             t, e = exc_info
             tb = None
 
+        for t in self.targets:
+            try:
+                t.fs.unlink(t.path)
+            except (IOError, OSError):
+                pass
+
         if t is None:
             # The Taskmaster didn't record an exception for this Task;
             # see if the sys module has one.
