Index: Action.py
===================================================================
*** Action.py	(revision 3802)
--- Action.py	(working copy)
***************
*** 504,515 ****
              l = string.join(self.presub_lines(env), '\n  ')
              out = "Building %s with action:\n  %s\n" % (t, l)
              sys.stdout.write(out)
!         s = None
          if show and self.strfunction:
!             s = self.strfunction(target, source, env)
!             if s:
                  if chdir:
!                     s = ('os.chdir(%s)\n' % repr(chdir)) + s
                  try:
                      get = env.get
                  except AttributeError:
--- 504,515 ----
              l = string.join(self.presub_lines(env), '\n  ')
              out = "Building %s with action:\n  %s\n" % (t, l)
              sys.stdout.write(out)
!         cmd = None
          if show and self.strfunction:
!             cmd = self.strfunction(target, source, env)
!             if cmd:
                  if chdir:
!                     cmd = ('os.chdir(%s)\n' % repr(chdir)) + cmd
                  try:
                      get = env.get
                  except AttributeError:
***************
*** 518,524 ****
                      print_func = get('PRINT_CMD_LINE_FUNC')
                      if not print_func:
                          print_func = self.print_cmd_line
!                 print_func(s, target, source, env)
          stat = 0
          if execute:
              if chdir:
--- 518,524 ----
                      print_func = get('PRINT_CMD_LINE_FUNC')
                      if not print_func:
                          print_func = self.print_cmd_line
!                 print_func(cmd, target, source, env)
          stat = 0
          if execute:
              if chdir:
***************
*** 536,542 ****
              finally:
                  if save_cwd:
                      os.chdir(save_cwd)
!         if s and save_cwd:
              print_func('os.chdir(%s)' % repr(save_cwd), target, source, env)
  
          return stat
--- 536,542 ----
              finally:
                  if save_cwd:
                      os.chdir(save_cwd)
!         if cmd and save_cwd:
              print_func('os.chdir(%s)' % repr(save_cwd), target, source, env)
  
          return stat
