Index: src/engine/SCons/Script/SConscript.py
===================================================================
RCS file: /cvsroot/scons/scons/src/engine/SCons/Script/SConscript.py,v
retrieving revision 1.109
diff -u -r1.109 SConscript.py
--- src/engine/SCons/Script/SConscript.py	12 Aug 2005 22:09:40 -0000	1.109
+++ src/engine/SCons/Script/SConscript.py	7 Jan 2006 04:25:54 -0000
@@ -149,7 +149,7 @@
 
 
 stack_bottom = '% Stack boTTom %' # hard to define a variable w/this name :)
-
+from os.path import abspath
 def _SConscript(fs, *files, **kw):
     top = fs.Top
     sd = fs.SConstruct_dir.rdir()
@@ -176,7 +176,7 @@
                 # fs match so we can open the SConscript.
                 fs.chdir(top, change_os_dir=1)
                 if f.rexists():
-                    _file_ = open(f.rstr(), "r")
+                    _file_ = open(abspath(f.rstr()), "r")
                 elif f.has_src_builder():
                     # The SConscript file apparently exists in a source
                     # code management system.  Build it, but then clear
@@ -185,7 +185,7 @@
                     f.build()
                     f.builder_set(None)
                     if f.exists():
-                        _file_ = open(str(f), "r")
+                        _file_ = open(abspath(str(f)), "r")
                 if _file_:
                     # Chdir to the SConscript directory.  Use a path
                     # name relative to the SConstruct file so that if
