Index: src/engine/SCons/Environment.py
===================================================================
RCS file: /cvs/scons/src/engine/SCons/Environment.py,v
retrieving revision 1.151
diff -u -r1.151 Environment.py
--- src/engine/SCons/Environment.py	21 Jan 2005 17:01:48 -0000	1.151
+++ src/engine/SCons/Environment.py	23 Jun 2005 00:10:10 -0000
@@ -564,6 +564,8 @@
         scanners.reverse()
         for scanner in scanners:
             for k in scanner.get_skeys(self):
+                if self['PLATFORM'] == 'win32':
+                    k = k.lower()
                 sm[k] = scanner
         return sm
         
@@ -572,6 +574,8 @@
         __cacheable__
         """
         sm = self._gsm()
+        if self['PLATFORM'] == 'win32':
+            # Value nodes have None skey
+            if skey != None:
+                skey = skey.lower()
         if sm.has_key(skey):
             return sm[skey]
         return None
