Index: Environment.py
===================================================================
--- Environment.py	(revision 2185)
+++ Environment.py	(working copy)
@@ -824,8 +824,27 @@
                         try:
                             add_to_orig = orig.append
                         except AttributeError:
-                            value.insert(0, orig)
-                            orig = value
+                            if key == 'CPPDEFINES':
+                                if SCons.Util.is_Dict(orig):
+                                    tmp = []
+                                    for i in orig.items():
+                                        if i[1] is not None:
+                                            tmp.append(i)
+                                        else:
+                                            tmp.append((i[0],))
+                                    orig = tmp
+                                if SCons.Util.is_Dict(value):
+                                    tmp = []
+                                    for i in value.items():
+                                        if i[1] is not None:
+                                            tmp.append(i)
+                                        else:
+                                            tmp.append((i[0],))
+                                    value = tmp
+                                orig = orig + value
+                            else:
+                                value.insert(0, orig)
+                                orig = value
                         else:
                             add_to_orig(value)
             t = []
