diff -ru SCons.orig/Tool/MSCommon/common.py SCons/Tool/MSCommon/common.py
--- SCons.orig/Tool/MSCommon/common.py	2009-09-20 01:00:01.000000000 +0200
+++ SCons/Tool/MSCommon/common.py	2009-12-26 12:43:55.000000000 +0100
@@ -31,6 +31,7 @@
 import os
 import subprocess
 import re
+import platform
 
 import SCons.Util
 
@@ -65,7 +66,7 @@
     # avoid repeated registry calls.
     global _is_win64
     if _is_win64 is None:
-        _is_win64 = has_reg(r"Software\Wow6432Node")
+        _is_win64 = (platform.machine() == "x64_32")
     return _is_win64
 
 
